File: packaging/copr/nabu-desktop-metas/generate-family-spec.py
#!/usr/bin/python3
"""Generate the unified desktop-meta spec from the five maintained profiles."""

from pathlib import Path
import re

HERE = Path(__file__).resolve().parent
SOURCE = HERE.parent / "nabu-unified-meta"
PROFILES = (
    "gnome-nabu-meta",
    "gnome-mobile-nabu-meta",
    "kde-plasma-nabu-meta",
    "kde-plasma-mobile-nabu-meta",
    "phosh-nabu-meta",
)
SOURCE_MAP = {
    "gnome-nabu-meta": {0: 0, 1: 1},
    "gnome-mobile-nabu-meta": {0: 0, 1: 1, 2: 13, 3: 14, 4: 15, 5: 16, 6: 17, 7: 18, 8: 19},
    "kde-plasma-nabu-meta": {0: 4, 1: 2, 2: 0, 3: 3, 4: 1, 5: 5, 6: 6, 7: 7, 8: 8},
    "kde-plasma-mobile-nabu-meta": {0: 9, 1: 10, 2: 11, 3: 12, 4: 2, 5: 0, 6: 3, 7: 1, 8: 5, 9: 6, 10: 7, 11: 8},
    "phosh-nabu-meta": {0: 0},
}


def text(name: str) -> str:
    return (SOURCE / f"{name}.spec").read_text()


def remap_sources(value: str, name: str) -> str:
    mapping = SOURCE_MAP[name]

    def replace(match: re.Match[str]) -> str:
        old = int(match.group(1))
        return f"%{{SOURCE{mapping[old]}}}"

    return re.sub(r"%\{SOURCE(\d+)\}", replace, value)


def section(value: str, heading: str) -> str:
    match = re.search(rf"(?ms)^%{re.escape(heading)}\s*\n(.*?)(?=^%(?:prep|build|install|check|files|post|preun|postun|posttrans|changelog)\b)", value)
    return match.group(1).rstrip() if match else ""


def description(value: str) -> str:
    match = re.search(r"(?ms)^%description\s*\n(.*?)(?=^%prep\b)", value)
    return match.group(1).strip()


def package_header(value: str, name: str) -> str:
    prefix = value.split("%description", 1)[0]
    lines = []
    for line in prefix.splitlines():
        if re.match(r"^(Summary|Requires|Recommends|Suggests|Conflicts|Provides|Obsoletes):", line):
            lines.append(line)
    return f"%package -n {name}\n" + "\n".join(lines)


def files(value: str, name: str) -> str:
    match = re.search(r"(?ms)^%files\s*\n(.*?)(?=^%(?:post|preun|postun|posttrans|changelog)\b)", value)
    if not match:
        raise RuntimeError("missing files section")
    result = match.group(1).rstrip()
    if name == "kde-plasma-nabu-meta":
        # Both translation archives carry the same SPDX license texts. Keep a
        # single packaged copy to avoid duplicate-file warnings.
        result = result.replace(" l10n/LICENSES/plasma-setup/*", "")
    return result


def scriptlets(value: str, name: str) -> str:
    matches = re.finditer(
        r"(?ms)^(%(?:pretrans|posttrans|preun|postun|pre|post))(?!\w)([^\n]*)\n(.*?)(?=^%(?:pretrans|posttrans|preun|postun|pre|post|files|changelog)\b)",
        value,
    )
    chunks = []
    for match in matches:
        command, options, body = match.groups()
        if " -n " not in f" {options} ":
            options = f" -n {name}{options}"
        chunks.append(f"{command}{options}\n{body.rstrip()}")
    return "\n\n".join(chunks)


headers = []
build_requires = set()
for profile in PROFILES:
    value = text(profile)
    headers.append(package_header(value, profile))
    for line in value.split("%description", 1)[0].splitlines():
        if line.startswith("BuildRequires:"):
            build_requires.add(line)

# GNOME Mobile is a strict superset of the GNOME payload. KDE Desktop carries
# the complete shared KDE payload; only the first mobile-session stanza is
# additional for Plasma Mobile. Avoid executing identical install operations
# twice, especially creation of fixed systemd symlinks.
gnome_install = remap_sources(section(text("gnome-mobile-nabu-meta"), "install"), "gnome-mobile-nabu-meta")
kde_install = remap_sources(section(text("kde-plasma-nabu-meta"), "install"), "kde-plasma-nabu-meta")
mobile_install = remap_sources(section(text("kde-plasma-mobile-nabu-meta"), "install"), "kde-plasma-mobile-nabu-meta")
mobile_session_install = mobile_install.split("install -Dm0755 kde-integration", 1)[0].rstrip()
install_parts = [
    "# Shared GNOME and GNOME Mobile payload\n" + gnome_install,
    "# Shared KDE Plasma payload\n" + kde_install,
    "# Plasma Mobile session-only payload\n" + mobile_session_install,
]
check_parts = []
for profile in ("gnome-mobile-nabu-meta", "kde-plasma-nabu-meta", "kde-plasma-mobile-nabu-meta"):
    check_parts.append(f"# {profile}\n" + remap_sources(section(text(profile), "check"), profile))

source_lines = """Source0:        nabu-kde-l10n-1.1.0.tar.gz
Source1:        nabu-flashlight-integration-1.0.0.tar.gz
Source2:        nabu-kde-integration-1.4.0.1.tar.gz
Source3:        nabu-kde-widgets-debug-1.0.1.tar.zst
Source4:        95-nabu-plasma-login.preset
Source5:        80-nabu-plasma-login-theme.conf
Source6:        nabu-plasma-login.svg
Source7:        90-nabu-powerdevil.conf
Source8:        90-nabu-compositor-realtime.conf
Source9:        plasma-mobile.desktop
Source10:       20-nabu-mobile-session.conf
Source11:       90-nabu-mobile-login.conf
Source12:       95-nabu-plasma-mobile.preset
Source13:       gnome-mobile-copr.repo
Source14:       nabu-gnome-mobile-sync
Source15:       nabu-gnome-mobile-sync.service
Source16:       nabu-gnome-mobile-sync.timer
Source17:       90-nabu-gnome-mobile-sync.preset
Source18:       test-gnome-mobile-repo-sync.sh
Source19:       20-nabu-mobile-user-mode.conf"""

out = [
    "%global debug_package %{nil}",
    "%global legacy_meta_max 9999999999-99",
    "",
    "Name:           nabu-desktop-metas",
    "Version:        3.0.0",
    "Release:        100%{?dist}",
    "Summary:        Unified desktop profile family for Xiaomi Pad 5",
    "License:        MIT AND GPL-2.0-or-later AND GPL-3.0-or-later AND BSD-2-Clause AND CC0-1.0",
    "URL:            https://github.com/MCC45TR/Nabu-Fedora-Rawhide-Builder",
    source_lines,
    "BuildArch:      noarch",
    *sorted(build_requires),
    "",
    "%description",
    "One source package for the five supported Nabu desktop manifests. Existing",
    "binary RPM names, dependencies, conflicts, integration payloads and update",
    "semantics are preserved for ordinary DNF upgrades.",
    "",
]

for profile, header in zip(PROFILES, headers):
    out.extend([header, "", f"%description -n {profile}", description(text(profile)), ""])

out.extend([
    "%prep",
    "%setup -q -c -T",
    "mkdir l10n flashlight kde-integration widgets",
    "tar -xzf %{SOURCE0} -C l10n --strip-components=1",
    "tar -xzf %{SOURCE1} -C flashlight --strip-components=1",
    "tar -xzf %{SOURCE2} -C kde-integration --strip-components=1",
    "tar --zstd -xf %{SOURCE3} -C widgets --strip-components=1",
    "chmod +x kde-integration/kde/senemos-nabu-color-profile kde-integration/tests/mock-kscreen-doctor",
    "",
    "%build",
    "",
    "%install",
    "\n\n".join(install_parts),
    "",
    "%check",
    "\n\n".join(check_parts),
    "",
])

for profile in PROFILES:
    scripts = scriptlets(text(profile), profile)
    if scripts:
        out.extend([scripts, ""])
    out.extend([f"%files -n {profile}", files(text(profile), profile), ""])

out.extend([
    "%changelog",
    "* Sun Sep 06 2026 mcc45tr <mcc45tr@gmail.com> - 3.0.0-100",
    "- Build all five desktop manifests from one COPR source family.",
    "- Preserve the existing binary names and stock Fedora/KDE package policy.",
    "",
])

(HERE / "nabu-desktop-metas.spec").write_text("\n".join(out))

File: packaging/copr/nabu-desktop-metas/generate-family-spec.cpp
// SPDX-License-Identifier: MIT
// Regenerate the reviewed unified COPR spec without a Python maintainer tool.
#include <algorithm>
#include <array>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <iterator>
#include <map>
#include <regex>
#include <set>
#include <stdexcept>
#include <string>
#include <vector>
#include <utility>

namespace fs = std::filesystem;

namespace {
using Lines = std::vector<std::string>;
const std::array<std::string, 5> profiles = {
    "gnome-nabu-meta", "gnome-mobile-nabu-meta", "kde-plasma-nabu-meta",
    "kde-plasma-mobile-nabu-meta", "phosh-nabu-meta"};
const std::map<std::string, std::map<int, int>> sourceMap = {
    {"gnome-nabu-meta", {{0, 0}, {1, 1}}},
    {"gnome-mobile-nabu-meta", {{0, 0}, {1, 1}, {2, 13}, {3, 14}, {4, 15},
                                 {5, 16}, {6, 17}, {7, 18}, {8, 19}}},
    {"kde-plasma-nabu-meta", {{0, 4}, {1, 2}, {2, 0}, {3, 3}, {4, 1},
                               {5, 5}, {6, 6}, {7, 7}, {8, 8}}},
    {"kde-plasma-mobile-nabu-meta", {{0, 9}, {1, 10}, {2, 11}, {3, 12},
                                      {4, 2}, {5, 0}, {6, 3}, {7, 1},
                                      {8, 5}, {9, 6}, {10, 7}, {11, 8}}},
    {"phosh-nabu-meta", {{0, 0}}}};

std::string read(const fs::path &path) {
    std::ifstream input(path, std::ios::binary);
    if (!input) throw std::runtime_error("cannot read " + path.string());
    return {std::istreambuf_iterator<char>(input), std::istreambuf_iterator<char>()};
}

Lines split(const std::string &value) {
    Lines result;
    std::string line;
    for (char ch : value) {
        if (ch == '\n') {
            result.push_back(line);
            line.clear();
        } else {
            line.push_back(ch);
        }
    }
    if (!line.empty()) result.push_back(line);
    return result;
}

std::string join(const Lines &value, const std::string &separator = "\n") {
    std::string result;
    bool first = true;
    for (const auto &line : value) {
        if (!first) result += separator;
        result += line;
        first = false;
    }
    return result;
}

std::string trimRight(std::string value) {
    while (!value.empty() && std::isspace(static_cast<unsigned char>(value.back())))
        value.pop_back();
    return value;
}

std::string trim(std::string value) {
    value = trimRight(std::move(value));
    const auto first = value.find_first_not_of(" \t\r\n\f\v");
    return first == std::string::npos ? "" : value.substr(first);
}

bool directive(const std::string &line, const std::string &name) {
    return line.rfind(name, 0) == 0 &&
           (line.size() == name.size() || std::isspace(static_cast<unsigned char>(line[name.size()])));
}

bool sectionEnd(const std::string &line) {
    static const std::array<std::string, 10> names = {
        "%prep", "%build", "%install", "%check", "%files", "%post",
        "%preun", "%postun", "%posttrans", "%changelog"};
    return std::any_of(names.begin(), names.end(), [&](const auto &name) { return directive(line, name); });
}

std::string section(const Lines &source, const std::string &heading) {
    Lines body;
    bool found = false;
    for (const auto &line : source) {
        if (!found) {
            found = line == "%" + heading;
        } else if (sectionEnd(line)) {
            break;
        } else {
            body.push_back(line);
        }
    }
    return found ? trimRight(join(body)) : "";
}

std::string description(const Lines &source) {
    Lines body;
    bool found = false;
    for (const auto &line : source) {
        if (!found) {
            found = line == "%description";
        } else if (directive(line, "%prep")) {
            break;
        } else {
            body.push_back(line);
        }
    }
    if (!found) throw std::runtime_error("missing description");
    return trim(join(body));
}

std::string packageHeader(const Lines &source, const std::string &name) {
    Lines result = {"%package -n " + name};
    static const std::regex relevant(R"(^(Summary|Requires|Recommends|Suggests|Conflicts|Provides|Obsoletes):)");
    for (const auto &line : source) {
        if (directive(line, "%description")) break;
        if (std::regex_search(line, relevant)) result.push_back(line);
    }
    return join(result);
}

std::string files(const Lines &source, const std::string &name) {
    Lines body;
    bool found = false;
    for (const auto &line : source) {
        if (!found) {
            found = line == "%files";
        } else if (directive(line, "%post") || directive(line, "%preun") ||
                   directive(line, "%postun") || directive(line, "%posttrans") ||
                   directive(line, "%changelog")) {
            break;
        } else {
            if (name == "kde-plasma-nabu-meta") {
                std::string adjusted = line;
                const std::string duplicate = " l10n/LICENSES/plasma-setup/*";
                const auto offset = adjusted.find(duplicate);
                if (offset != std::string::npos) adjusted.erase(offset, duplicate.size());
                body.push_back(std::move(adjusted));
            } else {
                body.push_back(line);
            }
        }
    }
    if (!found) throw std::runtime_error("missing files section: " + name);
    return trimRight(join(body));
}

bool scriptHeading(const std::string &line) {
    static const std::array<std::string, 6> names = {
        "%pretrans", "%posttrans", "%preun", "%postun", "%pre", "%post"};
    return std::any_of(names.begin(), names.end(), [&](const auto &name) { return directive(line, name); });
}

std::string scriptlets(const Lines &source, const std::string &name) {
    Lines chunks;
    for (std::size_t i = 0; i < source.size(); ++i) {
        if (!scriptHeading(source[i])) continue;
        const auto space = source[i].find_first_of(" \t");
        const std::string command = source[i].substr(0, space);
        std::string options = space == std::string::npos ? "" : source[i].substr(space);
        if (options.find(" -n ") == std::string::npos) options = " -n " + name + options;
        Lines body;
        for (++i; i < source.size() && !scriptHeading(source[i]) &&
                    !directive(source[i], "%files") && !directive(source[i], "%changelog"); ++i)
            body.push_back(source[i]);
        chunks.push_back(command + options + "\n" + trimRight(join(body)));
        if (i < source.size()) --i;
    }
    return join(chunks, "\n\n");
}

std::string remapSources(const std::string &value, const std::string &name) {
    static const std::regex source(R"(%\{SOURCE([0-9]+)\})");
    std::string result;
    std::sregex_iterator it(value.begin(), value.end(), source), end;
    std::size_t cursor = 0;
    for (; it != end; ++it) {
        result += value.substr(cursor, static_cast<std::size_t>(it->position()) - cursor);
        const int old = std::stoi((*it)[1].str());
        result += "%{SOURCE" + std::to_string(sourceMap.at(name).at(old)) + "}";
        cursor = static_cast<std::size_t>(it->position() + it->length());
    }
    return result + value.substr(cursor);
}

std::string generate(const fs::path &here) {
    const fs::path sourceDir = here.parent_path() / "nabu-unified-meta";
    std::map<std::string, Lines> sources;
    std::set<std::string> buildRequires;
    Lines headers;
    for (const auto &name : profiles) {
        auto lines = split(read(sourceDir / (name + ".spec")));
        headers.push_back(packageHeader(lines, name));
        for (const auto &line : lines) {
            if (directive(line, "%description")) break;
            if (line.rfind("BuildRequires:", 0) == 0) buildRequires.insert(line);
        }
        sources.emplace(name, std::move(lines));
    }

    const std::string previous = read(here / "nabu-desktop-metas.spec");
    const auto release = std::regex(R"((?:^|\n)Release:[ \t]*([^\n]+))");
    std::smatch releaseMatch;
    if (!std::regex_search(previous, releaseMatch, release))
        throw std::runtime_error("missing reviewed release in generated spec");
    const auto changelog = previous.find("%changelog\n");
    if (changelog == std::string::npos)
        throw std::runtime_error("missing reviewed changelog in generated spec");

    const auto &gnomeMobile = sources.at("gnome-mobile-nabu-meta");
    const auto &kde = sources.at("kde-plasma-nabu-meta");
    const auto &kdeMobile = sources.at("kde-plasma-mobile-nabu-meta");
    std::string mobileSession = remapSources(section(kdeMobile, "install"), "kde-plasma-mobile-nabu-meta");
    const auto marker = mobileSession.find("install -Dm0755 nabu-audio-orientation");
    if (marker == std::string::npos)
        throw std::runtime_error("mobile session marker missing; review generator before changing package layout");
    mobileSession = trimRight(mobileSession.substr(0, marker));

    Lines out = {
        "%global debug_package %{nil}", "%global legacy_meta_max 9999999999-99", "",
        "Name:           nabu-desktop-metas", "Version:        3.0.0",
        "Release:        " + releaseMatch[1].str(),
        "Summary:        Unified desktop profile family for Xiaomi Pad 5",
        "License:        MIT AND GPL-2.0-or-later AND GPL-3.0-or-later AND BSD-2-Clause AND CC0-1.0",
        "URL:            https://github.com/MCC45TR/Nabu-Fedora-Rawhide-Builder",
        R"(Source0:        nabu-kde-l10n-1.1.0.tar.gz
Source1:        nabu-flashlight-integration-1.0.0.tar.gz
Source2:        nabu-kde-integration-1.4.0.1.tar.gz
Source3:        nabu-kde-widgets-debug-1.0.1.tar.zst
Source4:        95-nabu-plasma-login.preset
Source5:        80-nabu-plasma-login-theme.conf
Source6:        nabu-plasma-login.svg
Source7:        90-nabu-powerdevil.conf
Source8:        90-nabu-compositor-realtime.conf
Source9:        plasma-mobile.desktop
Source10:       20-nabu-mobile-session.conf
Source11:       90-nabu-mobile-login.conf
Source12:       95-nabu-plasma-mobile.preset
Source13:       gnome-mobile-copr.repo
Source14:       nabu-gnome-mobile-sync
Source15:       nabu-gnome-mobile-sync.service
Source16:       nabu-gnome-mobile-sync.timer
Source17:       90-nabu-gnome-mobile-sync.preset
Source18:       test-gnome-mobile-repo-sync.sh
Source19:       20-nabu-mobile-user-mode.conf)"};
    out.insert(out.end(), buildRequires.begin(), buildRequires.end());
    out.insert(out.end(), {"", "%description", "One source package for the five supported Nabu desktop manifests. Existing",
                           "binary RPM names, dependencies, conflicts, integration payloads and update",
                           "semantics are preserved for ordinary DNF upgrades.", ""});
    for (std::size_t i = 0; i < profiles.size(); ++i) {
        const auto &name = profiles[i];
        out.insert(out.end(), {headers[i], "", "%description -n " + name,
                               description(sources.at(name)), ""});
    }
    out.insert(out.end(), {"%prep", "%setup -q -c -T", "mkdir l10n flashlight kde-integration widgets",
                           "tar -xzf %{SOURCE0} -C l10n --strip-components=1",
                           "tar -xzf %{SOURCE1} -C flashlight --strip-components=1",
                           "tar -xzf %{SOURCE2} -C kde-integration --strip-components=1",
                           "tar --zstd -xf %{SOURCE3} -C widgets --strip-components=1",
                           "chmod +x kde-integration/tests/mock-kscreen-doctor", "", "%build",
                           remapSources(section(kde, "build"), "kde-plasma-nabu-meta"), "", "%install",
                           "# Shared GNOME and GNOME Mobile payload\n" +
                               remapSources(section(gnomeMobile, "install"), "gnome-mobile-nabu-meta") +
                               "\n\n# Shared KDE Plasma payload\n" +
                               remapSources(section(kde, "install"), "kde-plasma-nabu-meta") +
                               "\n\n# Plasma Mobile session-only payload\n" + mobileSession,
                           "", "%check"});
    Lines checks;
    for (const auto &name : {"gnome-mobile-nabu-meta", "kde-plasma-nabu-meta", "kde-plasma-mobile-nabu-meta"})
        checks.push_back("# " + std::string(name) + "\n" +
                         remapSources(section(sources.at(name), "check"), name));
    out.insert(out.end(), {join(checks, "\n\n"), ""});
    for (const auto &name : profiles) {
        const auto scripts = scriptlets(sources.at(name), name);
        if (!scripts.empty()) out.insert(out.end(), {scripts, ""});
        out.insert(out.end(), {"%files -n " + name, files(sources.at(name), name), ""});
    }
    // Release and changelog are human-reviewed metadata. Keep them untouched.
    out.push_back(trimRight(previous.substr(changelog)));
    out.push_back("");
    return join(out);
}
} // namespace

int main(int argc, char **argv) {
    const bool checkOnly = argc == 3 && std::string(argv[1]) == "--check";
    if (!(argc == 2 || checkOnly)) {
        std::cerr << "usage: generate-family-spec [--check] SOURCE-DIRECTORY\n";
        return 2;
    }
    try {
        const fs::path here = fs::canonical(argv[checkOnly ? 2 : 1]);
        const std::string result = generate(here);
        if (checkOnly) {
            if (result != read(here / "nabu-desktop-metas.spec"))
                throw std::runtime_error("generated spec is stale; review source changes and regenerate");
            return 0;
        }
        std::ofstream output(here / "nabu-desktop-metas.spec", std::ios::binary | std::ios::trunc);
        if (!output) throw std::runtime_error("cannot write generated spec");
        output << result;
        if (!output) throw std::runtime_error("failed to write generated spec");
    } catch (const std::exception &error) {
        std::cerr << "generate-family-spec: " << error.what() << '\n';
        return 1;
    }
}
